File: examples/benchmarks/0_model_initializaiton/model_init.py
#!/usr/bin/env python
import timeit
import numpy as np
import pandas as pd

print("This script will benchmark eht-imaging's initialization of scattering models.")

# Scattering Model Name
model_labels = ["Dipole", "Von Mises", "Periodic Boxcar"]
model_names = ["dipole", "von_Mises", "boxcar"]

# Number of trials for timeit
ntrial = 1000  # this is a default number for @benchmark macro in Julia

# Common setup for timeit
setup = "import ehtim as eh"

# Looping for each scattering model
meanarr = []
stdarr = []
for i in range(3):
    print("  %d/3: %s Model" % (i + 1, model_labels[i]))
    model = model_names[i]

    # statement to be measured by timeit
    stmt = "eh.scattering.stochastic_optics.ScatteringModel(model='%s')" % (model)

    # measure time
    durations = timeit.Timer(stmt=stmt, setup=setup).repeat(repeat=ntrial, number=1)

    # printout the mean and standard deviation
    print("%.2f +/- %.2f ms" % (np.mean(durations) * 1e3, np.std(durations) * 1e3))

    meanarr.append(np.mean(durations) * 1e3)
    stdarr.append(np.std(durations) * 1e3)

df = pd.DataFrame(dict(model=model_names, mean_ms=meanarr, std_ms=stdarr))
df.to_csv("ehtim_model_init.csv")

File: examples/benchmarks/0_model_initializaiton/model_init.jl
using Pkg
Pkg.activate(dirname(@__DIR__))

using BenchmarkTools
using CSV
using DataFrames
using Printf
using ScatteringOptics

# the directory above the script directory
model_names = ["Dipole", "Von Mises", "Periodic Boxcar"]
model_names_ehtim = ["dipole", "von_Mises", "boxcar"]
model_names_so = [DipoleScatteringModel, vonMisesScatteringModel, PeriodicBoxCarScatteringModel]

@info "This script will benchmark the initialization time of the scattering models, critical for the scattering parameter inference"

@info "Step 1: Benchmarking ScatteringOptics.jl"
so_mean_time = []
so_std_time = []
so_label = []
for i in 1:3
    @info "  $(i)/3: $(model_names[i]) Model"
    local model = model_names_so[i]
    local out = @benchmark $model()
    push!(so_mean_time, mean(out).time/1e6)
    push!(so_std_time, std(out).time/1e6)
    local label = @sprintf("%.2f +/- %.2f ms",mean(out).time/1e6, std(out).time/1e6)
    push!(so_label, label)
end

@info "Step 2: Loading results from eht-imaging library"
ehdf = DataFrame(CSV.File("./ehtim_model_init.csv"))
eh_label = []
for i in 1:3
    @info "  $(i)/3: $(model_names[i]) Model"
    local meanval = ehdf[i, :mean_ms]
    local stdval = ehdf[i, :std_ms]
    local label = @sprintf("%.2f +/- %.2f ms", meanval, stdval)
    push!(eh_label, label)
end

@info "Here is the results of the benchmarking"
df = DataFrame(
    [(Name="ScatteringOptics.jl", Dipole=so_label[1], vonMises=so_label[2], PeriodicBoxCar=so_label[3]),
    (Name="eht-imaging (Python)", Dipole=eh_label[1], vonMises=eh_label[2], PeriodicBoxCar=eh_label[3])]
)
print(df)
open("results.txt","w") do io
    println(io, df)
end